Add -q option to `docker service ps`

gaocegege opened the pull request:
**\- What I did**

fixes #27643

**\- How to verify it**

```
$ docker swarm init
$ docker service create --name redis --replicas=5 redis:3.0.6
$ docker service ps -q redis
```

**\- A picture of a cute animal (not mandatory but encouraged)**

<img src="https://octodex.github.com/images/front-end-conftocat.png" width="100">

gaocegege commented on cli/command/task/print.go:
```diff
@@ -93,18 +114,21 @@ func Print(dockerCli *command.DockerCli, ctx context.Context, tasks []swarm.Task
-			taskErr,
-		)
+		if quiet {
+			fmt.Fprintln(out, indentedName)
```
>In addition, if quiet, maybe we do not have to process the err things. Just print when constructing of indentedName finishes. 😄

🤔 Yeah, there is no need to process errs.

>I think `indentedName = fmt.Sprintf(" \\_ %s", indentedName)` above may be displayed via this line of code. If that, I have no idea if it is reasonable. WDYT @gaocegege

Ah, okay I think we should pass the indent logic when `ps -q`

```
		if quiet {
			fmt.Fprintln(out, indentedName)
			continue
		}
		// Indent the name if necessary
		indentedName := name
		logrus.Debugf("DEBUG: %d, %d", prevSlot, task.Slot)
		// Since the new format of the task name is <ServiceName>.<Slot>.<taskID>, we should only compare
		// <ServiceName> and <Slot> here.
		if prevServiceName == serviceName {
			indentedName = fmt.Sprintf(" \\_ %s", indentedName)
		}
		prevServiceName = serviceName
		prevSlot = task.Slot
``` 

Then the result would be 

```diff
redis.1.mwvj0azq00lcydntsv79uj2k6
-  \_ redis.1.zk9fhepa7cgnoex3pgn0sngir
+ redis.1.zk9fhepa7cgnoex3pgn0sngir
```
```

Or do you have a better idea about indented name?

allencloud commented:
>
We need a -q option that removes headers and only does ID.

I am afraid `redis.1.` in `redis.1.6u87kp12lrw1ikrejqti51g8y` is not included in task id. @gaocegege

gaocegege commented:
PTAL

```
$ docker service ps redis
NAME                               IMAGE        NODE          DESIRED STATE  CURRENT STATE            ERROR
redis.1.xrlbr5skxm5mfquugv7igtw21  redis:3.0.6  56a8544282d8  Running        Preparing 9 seconds ago
redis.2.69bi767slr5fmpb37n78zop4m  redis:3.0.6  56a8544282d8  Running        Preparing 9 seconds ago
$ docker service ps -q redis
xrlbr5skxm5mfquugv7igtw21
69bi767slr5fmpb37n78zop4m
```

LK4D4 commented:
LGTM
I'm not sure what to do with docs, though. ping @thaJeztah

thaJeztah commented:
> I'm not sure what to do with docs, though. 

This would only need changes to the reference docs (man-page for this is generated); @gaocegege can you update https://github.com/docker/docker/blob/master/docs/reference/commandline/service_ps.md ?


It also needs a change to the bash and zsh completion scripts, so /cc @albers @sdurrheimer

gaocegege commented:
re @thaJeztah 

I have updated the docs. And I found there is no `-a` option now, so I removed it and add `-q`.

[preview](https://github.com/gaocegege/docker/blob/636d2bfb2d46eae552746c8b973e94a0a8a08525/docs/reference/commandline/service_ps.md)

sdurrheimer commented:
zsh completion LGTM

albers commented on contrib/completion/bash/docker:
```diff
@@ -2462,7 +2462,7 @@ _docker_service_ps() {
 	case "$cur" in
 		-*)
-			COMPREPLY=( $( compgen -W "--all -a --filter -f --help --no-resolve --no-trunc" -- "$cur" ) )
+			COMPREPLY=( $( compgen -W "--quiet -q --filter -f --help --no-resolve --no-trunc" -- "$cur" ) )
```
Please preserve alphabetical order here.

gaocegege commented on contrib/completion/zsh/_docker:
```diff
@@ -1167,7 +1167,7 @@ __docker_service_subcommand() {
             _arguments $(__docker_arguments) \
                 $opts_help \
-                "($help -a --all)"{-a,--all}"[Display all tasks]" \
+                "($help -q --quiet)"{-q,--quiet}"[Only display IDs]" \
```
OK

albers reviewed (approved):
bash completion LGTM, Thanks!

sdurrheimer reviewed (approved):
re-LGTM

vdemeester reviewed (approved):
LGTM 👼
/cc @aaronlehmann @stevvooe

stevvooe commented on cli/command/task/print.go:
```diff
@@ -50,6 +52,27 @@ func Print(dockerCli *command.DockerCli, ctx context.Context, tasks []swarm.Task
+}
+
+// PrintQuiet shows task list in a quiet way.
+func PrintQuiet(dockerCli *command.DockerCli, tasks []swarm.Task) error {
```
No need to export this.

stevvooe replied:
I see: this package is useless.

Just move `print.go` to the service package. This isn't used elsewhere.

stevvooe replied:
No reason to preserve the package import. These functions are a for loop and println.

Perhaps, this is outside the scope of this pr. Let's leave as is and fix in another pr.

stevvooe reviewed (changes requested):
Please don't export functions that don't need to be exported.

stevvooe commented:
LGTM

If someone wants de-leftpad the task package, that would be awesome!

vdemeester commented:
Moving this to docs-review then 👼

thaJeztah commented on cli/command/service/ps.go:
```diff
@@ -32,6 +33,7 @@ func newPsCommand(dockerCli *command.DockerCli) *cobra.Command {
 		},
 	}
 	flags := cmd.Flags()
+	flags.BoolVarP(&opts.quiet, "quiet", "q", false, "Only display IDs")
```
Looks good, wondering if we should use `Only display task IDs` (there's so many ID's related to this overview, it became a bit confusing to me)

@vdemeester wdyt?

vdemeester replied:
ah right, I do think `Only display task IDs` would make it a little bit clearer indeed 👼

gaocegege replied:
Ah, OK

I would add it.

cpuguy83 commented:
Can you squash commits, please?

thaJeztah commented:
LGTM after commits are squashed

gaocegege committed:
fixes #27643

Signed-off-by: Ce Gao <ce.gao@outlook.com>

vdemeester reviewed (approved):
LGTM 🐸
